protocols/bizcoin/bizcoin_simulation.go

```diff
@@ -0,0 +1,166 @@
+}
+
+// Simulation implements da.Simulation interface
+type Simulation struct {
+	// sda fields:
+	sda.SimulationBFTree
+	// your simulation specific fields:
+	simulationConfig
+}
+
+type simulationConfig struct {
+	// Blocksize is the number of transactions in one block:
+	Blocksize int
+	//blocksDir is the directory where to find the transaction blocks (.dat files)
+	BlocksDir string
+	// timeout the leader after TimeoutMs milliseconds
+	TimeoutMs uint64
+	// Fail:
+	// 0  do not fail
+	// 1 fail by doing nothing
+	// 2 fail by sending wrong blocks
+	Fail uint
+}
+
+func NewSimulation(config string) (sda.Simulation, error) {
+	es := &Simulation{}
+	_, err := toml.Decode(config, es)
+	if err != nil {
+		return nil, err
+	}
+	return es, nil
+}
+
+// Setup implements sda.Simulation interface
+func (e *Simulation) Setup(dir string, hosts []string) (*sda.SimulationConfig, error) {
+	sc := &sda.SimulationConfig{}
+	e.CreateEntityList(sc, hosts, 2000)
+	err := e.CreateTree(sc)
+	if err != nil {
+		return nil, err
+	}
+	return sc, nil
+}
+
+type monitorMut struct {
+	*monitor.Measure
+	sync.Mutex
+}
+
+func (m *monitorMut) NewMeasure(id string) {
+	m.Lock()
+	defer m.Unlock()
+	m.Measure = monitor.NewMeasure(id)
+}
+func (m *monitorMut) MeasureAndReset() {
+	m.Lock()
+	defer m.Unlock()
+	m.Measure = nil
+}
+
+// Run implements sda.Simulation interface
+func (e *Simulation) Run(sdaConf *sda.SimulationConfig) error {
+	dbg.Lvl1("Simulation starting with:  Rounds=", e.Rounds)
+	server := NewBizCoinServer(e.Blocksize, e.TimeoutMs, e.Fail)
+
+	node, _ := sdaConf.Overlay.NewNodeEmptyName("Broadcast", sdaConf.Tree)
+	proto, _ := broadcast.NewBroadcastRootProtocol(node)
+	node.SetProtocolInstance(proto)
+	// channel to notify we are done
+	broadDone := make(chan bool)
+	proto.RegisterOnDone(func() {
+		broadDone <- true
+	})
+	proto.Start()
+	// wait
+	<-broadDone
+
+	for round := 0; round < e.Rounds; round++ {
+		client := NewClient(server)
+		client.StartClientSimulation(e.BlocksDir, e.Blocksize)
+
+		dbg.Lvl1("Starting round", round)
+		// create an empty node
+		node, err := sdaConf.Overlay.NewNodeEmptyName("BizCoin", sdaConf.Tree)
+		if err != nil {
+			return err
+		}
+		// instantiate a bizcoin protocol
+		rComplete := monitor.NewMeasure("round")
+		pi, err := server.Instantiate(node)
+		if err != nil {
+			return err
+		}
+
+		bz := pi.(*BizCoin)
+		// Register callback for the generation of the signature !
+		bz.RegisterOnSignatureDone(func(sig *BlockSignature) {
+			rComplete.Measure()
+			if err := verifyBlockSignature(node.Suite(), node.EntityList().Aggregate, sig); err != nil {
+				dbg.Lvl1("Round", round, " FAILED:", err)
+			} else {
+				dbg.Lvl1("Round", round, " SUCCESS")
+			}
+		})
+
+		// Register when the protocol is finished (all the nodes have finished)
+		done := make(chan bool)
+		bz.RegisterOnDone(func() {
+			dbg.Print("SIMULATION ON DONE CALLED")
```

Comment:
`dbg.Print` is only to be used for direct debugging and then to be taken away. No allcaps, please.

And: this callback-function could definitively gain by being replaced by a channel.